junit-platform-commons/src/main/java/org/junit/platform/commons/util/ClasspathScanner.java

```diff
@@ -154,6 +203,22 @@ private void processClassFileSafely(Path baseDir, String basePackageName, ClassF
 		}
 	}
 
+	private void processResourceFileSafely(Path baseDir, String basePackageName, Predicate<Resource> resourceFilter,
+			Path resourceFile, Consumer<Resource> resourceConsumer) {
+		try {
+			String fullyQualifiedResourceName = determineFullyQualifiedResourceName(baseDir, basePackageName,
+				resourceFile);
+			Resource resource = new ClasspathResource(fullyQualifiedResourceName, resourceFile.toUri());
```

Comment:
This might be a problem.

1. Resources (and classes) with the same name can appear multiple times on the classpath.
2. So while walking the file tree, we can encounter the same resource twice
3. For classes the class loader ensures the same class is loaded in `processClassFileSafely` and eventually these duplicates are deduped by `.distinct()`. But in `processResourceFileSafely` we do not have this.
4. So we can end up with resource `<classpath-root-a>/com/example/resource.txt` and `<classpath-root-b>/com/example/resource.txt`.
5. This does not match the behaviour of `ClassLoader.getResource()`.

Reply:
Do you think it would it be sufficient to document that multiple resources with the same name (but different URLs) may be returned?

Reply:
No. I don't think that would be sufficient. 

I would expect the behavior of the `ClasspathResourceSelector` for a given resource name to be consistent with `ClassLoader.getResource()`. And because the `ResourceContainerSelectorResolver` produces class path resource selectors with a resource through `ReflectionUtils.findAllResourcesIn*` the resources discovered this way should also be consistent with `ClassLoader.getResource()`.

In essence given this structure:
``` 
<classpath-root-a>
  - com/example/resource.txt 
 <classpath-root-b>
 - com/example/resource.txt.
```

Then using either
 * `DiscoverySelectors.selectPackage("com.example")`
 * `DiscoverySelectors.selectResource("com/example/resource.txt")`

Should result in a single test derived from the resource with `ClassLoader.getResource("com/example/resource.txt").toUri()`. As is, using the package selector would result in two tests.

Reply:
I'll try to add some tests for this. Shadowing resources on the class path is not obvious.

Reply:
I agree that in most cases, a single resource should be returned. I don't have a concrete use case for finding all of them, i.e. in different classpath roots. It came up recently in the context of potentially supporting multiple `junit-platform.properties` files.